Docs - Cms Kit Blogs (#8861)
abpframework/abp: docs/en/Modules/cms-kit/Blog-System.md

# Blog System

> CMS kit provides a **Blog** system to to add Blogs & Posts to public website. This section describes the details of the Blog system.

Reviewer: ```suggestion
CMS kit provides a **blog** system to add blogs & posts to a public website. This section describes the details of the blog system.
```

## Feature

> CMS kit uses the [global feature](https://docs.abp.io/en/abp/latest/Global-Features) system for all implemented features. Commercial startup templates come with all the CMS Kit related features are enabled by default, if you create the solution with the public website option. If you're installing the CMS Kit manually or want to enable the rating feature individually, open the `GlobalFeatureConfigurator` class in the `Domain.Shared` project and place the following code to the `Configure `method.

Reviewer: ```suggestion
CMS kit uses the [global feature](https://docs.abp.io/en/abp/latest/Global-Features) system for all implemented features. Commercial startup templates come with all the CMS Kit-related features are enabled by default, if you create the solution with the public website option. If you're installing the CMS Kit manually or want to enable the blog feature individually, open the `GlobalFeatureConfigurator` class in the `Domain.Shared` project and place the following code to the `Configure `method.
```

```csharp
GlobalFeatureManager.Instance.Modules.CmsKit(cmsKit =>
{
	cmsKit.Blogs.Enable();
});
```

# Internals

## Domain Layer

#### Aggregates

This module follows the [Entity Best Practices & Conventions](https://docs.abp.io/en/abp/latest/Best-Practices/Entities) guide.

- `Blog` _(aggregate root)_: Presents blogs of application.

> - `BlogPost`_(aggregate root)_: Presents Blog Posts by Blogs

Reviewer: ```suggestion
- `BlogPost`_(aggregate root)_: Presents blog posts in blogs.
```

> - `BlogFeauture`:_(aggregate root)_: Presents blog features enabled/disabled state. Such as reactions, ratings, comments etc.

Reviewer: ```suggestion
- `BlogFeature`:_(aggregate root)_: Presents blog features enabled/disabled state. Such as reactions, ratings, comments, etc.
```

#### Repositories

This module follows the [Repository Best Practices & Conventions](https://docs.abp.io/en/abp/latest/Best-Practices/Repositories) guide.

Following custom repositories are defined for this feature:

- `IBlogRepository`
- `IBlogPostRepository`
- `IBlogFeatureRepository`

#### Domain services

This module follows the [Domain Services Best Practices & Conventions](https://docs.abp.io/en/abp/latest/Best-Practices/Domain-Services) guide.

- `BlogManager`: Includes some operations for `Blog` aggreate root to keep data consistency.

>   - `CreateAsync`: Creates a new Blog entity and makes sure slug is used one time.

Reviewer: ```suggestion
  - `CreateAsync`: Creates a new `Blog` entity and makes sure slug is used one time.
```

>   - `UpdateAsync`: Updates existing Blog entity and makes sure slug is used one time.

Reviewer: ```suggestion
  - `UpdateAsync`: Updates existing `Blog` entity and makes sure slug is used one time.
```

- `BlogPostManager`: Includes some operations for `BlogPost` aggreagate root such as creating & updating.
  - `CreateAsync`: Creates a new BlogPost and makes sure slug is used only one time.
  - `SetSlugUrlAsync`: Sets `UrlSlug` property of BlogPost entity and makes sure slug isn't duplicated in same blog.
- `BlogFeatureManager`: Includes some operations for managing blog features.
  - `SetAsync`: Sets a feature enabled/disabled for a Blog.
  - `SetDefaultsAsync`: Resets all feature configuration for a Blog.

### Application layer

#### Application Services

##### Admin

> - `BlogAdminAppService` _(implements `IBlogAdminAppService`)_: Implements use cases of Blog Management.

Reviewer: ```suggestion
- `BlogAdminAppService` _(implements `IBlogAdminAppService`)_: Implements use cases of blog management.
```

> - `BlogFeatureAdminAppService` _(implements `IBlogFeatureAdminAppService`)_: Implements use cases of Blog Feature Management.

Reviewer: ```suggestion
- `BlogFeatureAdminAppService` _(implements `IBlogFeatureAdminAppService`)_: Implements use cases of blog feature management.
```

> - `BlogPostAdminAppService` _(implements `IBlogPostAdminAppService`)_: Implements use cases of Blog Post Management.

Reviewer: ```suggestion
- `BlogPostAdminAppService` _(implements `IBlogPostAdminAppService`)_: Implements use cases of blog post management.
```

##### Common

- `BlogFeatureAppService` _(implements I`BlogFeatureAppService`)_: Implements feature checking operations.

##### Public

> - `BlogPostPublicAppService` _(implements `IBlogPostPublicAppService`)_: Implements public blog post displaying operations.

Reviewer: ```suggestion
- `BlogPostPublicAppService` _(implements `IBlogPostPublicAppService`)_: Implements use cases of blog posts management on public websites.
```
